internal/service/common/async/store.go

```diff
@@ -0,0 +1,216 @@
+type AsyncEventType int
+
+const (
+	// Updated indicates that the object is inserted or updated
+	Updated AsyncEventType = iota
+	// Deleted indicates that the object is deleted
+	Deleted
+	// SyncComplete indicates that the Reflector has re-listed all objects from the API server
+	SyncComplete
+)
+
+// AsyncChangeEvent defines the data received by a data source to signal that an async handleWatchEvent has been received for a
+// given object type.
+type AsyncChangeEvent struct {
+	DataSourceID uuid.UUID
+	EventType    AsyncEventType
+	Object       db.Model
+	Keys         []uuid.UUID
+}
+
+// AsyncEventHandler is intended to be implemented by the Collector so that it can receive data from the Reflector via
+// the Store interface.
+type AsyncEventHandler interface {
+	// HandleSyncComplete is used to indicate that the full list of objects has been received.  Any stale objects should now
+	// be deleted.
+	HandleSyncComplete(ctx context.Context, objectType runtime.Object, keys []uuid.UUID) error
+	// HandleAsyncEvent is used to pass objects received from the Reflector to the Collector
+	HandleAsyncEvent(ctx context.Context, obj interface{}, eventType AsyncEventType) (uuid.UUID, error)
+}
+
+// operation defines an internal structure used to pass information around about pending operations.
+type operation struct {
+	eventType AsyncEventType
+	// objects is used to store the subject(s) of the operation.  If eventType is Updated or Deleted,
+	// then this slice will contain only a single element.  If eventType is SyncComplete, then it
+	// may contain 0 or more elements.
+	objects []interface{}
+}
+
+// ReflectorStore defines an adaptation layer between a Reflector and our Collector.  This is an alternative to using an
+// Informer which eliminates the need for caching objects in memory.  On a system engineered to capacity, this can save
+// us hundreds of MB of memory.
+type ReflectorStore struct {
+	// ObjectType is an instance of the object type monitored by the Reflector.
+	ObjectType runtime.Object
+	// hasSynced indicates whether the initial sync has completed (i.e., whether Replace has been called)
+	hasSynced bool
+	// queue stores the pending operations received from the Reflector
+	queue []operation
+	// mutex protects the shared instance variables from concurrency issues
+	mutex sync.Mutex
+	// ready is used to signal that a new operation has been received from the Reflector
+	ready chan struct{}
+}
+
+// NewReflectorStore creates a new ReflectorStore
+func NewReflectorStore(objectType runtime.Object) *ReflectorStore {
+	return &ReflectorStore{
+		ObjectType: objectType,
+		queue:      []operation{},
+		ready:      make(chan struct{}, 1),
+	}
+}
+
+// enqueue adds an operation to the queue and signals the receiver if it is the first operation
+// enqueued to the queue.
+func (c *ReflectorStore) enqueue(operation operation) {
+	c.mutex.Lock()
+	defer c.mutex.Unlock()
+	c.queue = append(c.queue, operation)
```

Comment:
Wonder if it would benefit if we limit the size of the q. Like block until maxQueueSize for when operations are being added faster than they can be processed.

Reply:
I don't expect for this to get this out of hand given that there shouldn't be more than a few thousand in the worse case which is startup.   We could add a high-watermark type log here to track what the behaviour is first and then add some sort of rate limiting if we see it exceed what we think is 'normal'

Reply:
Added a log if the high watermark jumps by 10 (to avoid excessive logs)